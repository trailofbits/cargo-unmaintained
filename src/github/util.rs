use anyhow::{Context, Result, anyhow};
use std::{
    env::var,
    fs::{File, OpenOptions, create_dir_all, read_to_string},
    io::{Write, stdin},
    path::PathBuf,
    sync::{LazyLock, OnceLock},
};

#[allow(clippy::unwrap_used)]
static CONFIG_DIRECTORY: LazyLock<PathBuf> = LazyLock::new(|| {
    #[cfg(not(windows))]
    {
        let base_directories = xdg::BaseDirectories::new().unwrap();
        base_directories.get_config_file("cargo-unmaintained")
    }
    #[cfg(windows)]
    {
        let local_app_data = var("LOCALAPPDATA").unwrap();
        PathBuf::from(local_app_data).join("cargo-unmaintained")
    }
});

static TOKEN_PATH: LazyLock<PathBuf> = LazyLock::new(|| CONFIG_DIRECTORY.join("token.txt"));

pub(super) static PERSONAL_TOKEN: OnceLock<String> = OnceLock::new();

pub fn load_token(f: impl FnOnce(&str) -> Result<()>) -> Result<bool> {
    let token_untrimmed = if let Ok(path) = var("GITHUB_TOKEN_PATH") {
        read_to_string(&path).with_context(|| format!("failed to read {path:?}"))?
    } else if let Ok(token) = var("GITHUB_TOKEN") {
        // smoelius: Suppress warning if `CI` is set, i.e., if running on GitHub.
        if var("CI").is_err() {
            #[cfg(__warnings)]
            crate::warn!(
                "found a token in `GITHUB_TOKEN`; consider using the more secure method of \
                 setting `GITHUB_TOKEN_PATH` to the path of a file containing the token",
            );
        }
        token
    } else if TOKEN_PATH.try_exists().with_context(|| {
        format!(
            "failed to determine whether `{}` exists",
            TOKEN_PATH.display()
        )
    })? {
        read_to_string(&*TOKEN_PATH).with_context(|| format!("failed to read {TOKEN_PATH:?}"))?
    } else {
        #[cfg(__warnings)]
        crate::warn!(
            "`GITHUB_TOKEN_PATH` and `GITHUB_TOKEN` are not set and no file was found at {}; \
             archival statuses will not be checked",
            TOKEN_PATH.display()
        );
        return Ok(false);
    };
    let token = token_untrimmed.trim_end().to_owned();
    PERSONAL_TOKEN
        .set(token.clone())
        .map_err(|_| anyhow!("`load_token` was already called"))?;
    f(&token)?;
    Ok(true)
}

pub(crate) fn save_token() -> Result<()> {
    println!("Please paste a personal access token below. The token needs no scopes.");

    let mut buf = String::new();

    {
        let n = stdin()
            .read_line(&mut buf)
            .with_context(|| "failed to read stdin")?;
        assert_eq!(buf.len(), n);
    }

    create_dir_all(&*CONFIG_DIRECTORY).with_context(|| "failed to create config directory")?;

    let mut file = OpenOptions::new()
        .create(true)
        .truncate(true)
        .write(true)
        .open(&*TOKEN_PATH)
        .with_context(|| format!("failed to open `{}`", TOKEN_PATH.display()))?;
    set_permissions(&file, 0o600)?;
    file.write_all(buf.as_bytes())
        .with_context(|| format!("failed to write `{}`", TOKEN_PATH.display()))?;

    println!(
        "Personal access token written to `{}`",
        TOKEN_PATH.display()
    );

    Ok(())
}

type CargoResult<T> = Result<T>;

// smoelius: The below definitions of `set_permissions` were copied from:
// https://github.com/rust-lang/cargo/blob/1e6828485eea0f550ed7be46ef96107b46aeb162/src/cargo/util/config.rs#L1010-L1024
#[cfg(unix)]
#[cfg_attr(dylint_lib = "try_io_result", allow(try_io_result))]
fn set_permissions(file: &File, mode: u32) -> CargoResult<()> {
    use std::os::unix::fs::PermissionsExt;

    let mut perms = file.metadata()?.permissions();
    perms.set_mode(mode);
    file.set_permissions(perms)?;
    Ok(())
}

#[cfg(not(unix))]
#[allow(unused, clippy::unnecessary_wraps)]
fn set_permissions(file: &File, mode: u32) -> CargoResult<()> {
    Ok(())
}
