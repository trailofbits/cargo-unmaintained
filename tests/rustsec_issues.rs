use snapbox::{Data, assert_data_eq};
use std::{
    env::{remove_var, var},
    fs::write,
    path::PathBuf,
    process::Command,
};

mod util;
use util::{Tee, tee};

#[ctor::ctor]
fn initialize() {
    remove_var("CARGO_TERM_COLOR");
}

#[test]
fn rustsec_issues() {
    const PATH_STDOUT: &str = "tests/rustsec_issues.stdout";

    let mut command = Command::new("cargo");
    command.args(["run", "--example=rustsec_issues"]);

    let output = tee(command, Tee::Stdout).unwrap();

    let stdout_actual = std::str::from_utf8(&output.captured).unwrap();

    if var("BLESS").is_ok() {
        write(PATH_STDOUT, stdout_actual).unwrap();
    } else {
        assert_data_eq!(
            stdout_actual,
            Data::read_from(&PathBuf::from(PATH_STDOUT), None),
        );
    }
}
