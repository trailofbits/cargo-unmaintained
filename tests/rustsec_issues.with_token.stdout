term...not found
ignoring `https://twitter.com/NaolDuga/status/1166705826192678914`
ignoring `https://twitter.com/bascule/status/1166567763411947520`
chan...found
termcolor...not found
rust-crypto...found
ignoring `https://github.com/RustCrypto`
term...not found
rusb...not found
ring...not found
crossbeam...not found
libusb-rs...not found
sodiumoxide...found
ignoring `https://internals.rust-lang.org/t/tracking-unmaintained-crates-using-rustsec/10884`
chan...found
rusb...not found
libusb-rs...not found
rust-crypto...found
term...not found
cassandra-rs...not found
quickersort...not found
rust_sodium...found
xdg...not found
home...not found
directories-rs...not found
dirs-rs...not found
dirs-sys-rs...not found
ignoring `https://www.reddit.com/r/rust/comments/ga7f56/why_dirs_and_directories_repositories_have_been/`
ignoring `https://www.reddit.com/r/rust/comments/ga7f56/why_dirs_and_directories_repositories_have_been/foyqipp/`
traits...not found
failure...found
ignoring `https://internals.rust-lang.org/t/failure-crate-maintenance/12087`
stderr...found
rust-xcb...not found
office...found
crate...not found
calamine...not found
rulinalg...found
ab-glyph...not found
ignoring `https://gitlab.redox-os.org/redox-os/rusttype/-/issues/148`
block-ciphers...found
block-ciphers...found
block-ciphers...found
utils...found
anymap...not found
cosmos-rust...not found
destructure_traitobject...not found
rust-traitobject...not found
dryoc...not found
AEADs...not found
nacl-compat...found
libsodium-sys-stable...not found
sodiumoxide...found
ignoring `https://speakerdeck.com/tarcieri/thoughts-on-rust-cryptography?slide=84`
rust-ansi-term...not found
windows-rs...not found
serde_cbor...found
cbor...found
xml-rs...not found
cargo-download...not found
ignoring `https://user-images.githubusercontent.com/19519553/147587813-34afbc8e-cee9-4101-972b-f8e4b9f511e9.png`
ignoring `https://user-images.githubusercontent.com/19519553/147587813-34afbc8e-cee9-4101-972b-f8e4b9f511e9.png`
lmdb-rkv...found
lmdb-rs...found
prettytable-rs...not found
owning-ref-unsoundness...not found
futures-intrusive...not found
users...not found
ignoring `https://github.com/dtolnay?tab=repositories&type=archived,`
markdown...not found
ignoring `https://doc.rust-lang.org/std/vec/struct.Vec.html#method.from_raw_parts`
maligned...not found
chrono...not found
toml-rs...not found
sass-rs...found
grass...not found
rsass...not found
static_type_map...found
erased_set...not found
double-checked-cell...not found
dotenv...not found
anymap...not found
ignoring `https://rustsec.org/advisories/RUSTSEC-2021-0065`
interledger-rs...not found
serde-xml-rs...not found
interledger-rs...not found
anymap...not found
wee_alloc...found
ignoring `https://www.reddit.com/r/rust/comments/x1cle0/dont_use_wee_alloc_in_production_code_targeting/`
badge...found
docs...not found
num-format...not found
ttmap...not found
typemap...not found
typemap_rev...not found
rust-typemap...not found
num-format...not found
orbtk...found
multipart...found
parity-wasm...found
wyz...not found
http-types...found
surf...not found
hyper...not found
rusoto...found
rustc-hex...not found
rust-claim...not found
json...not found
json-rust...not found
daemonize...not found
buf_redux...found
rust-boxfnonce...not found
mach2...not found
mach...not found
rust-encoding...not found
tui-rs...not found
tui-rs-revival...not found
kuchiki...found
buf_redux...found
wither...found
mbox...not found
crypto-hash...found
malloc_buf...not found
AEADs...not found
sysinfo...not found
rust-users...not found
io_ext...found
cloud-storage...found
cloud-storage-rs...found
dlopen2...not found
quote...not found
syn...not found
teams...not found
ignoring `https://github.com/rust-bus`
hpack...found
hpack-patched...not found
ignoring `https://doc.rust-lang.org/cargo/reference/overriding-dependencies.html`
loopdev-erikh...found
loopdev...not found
rust-cpython...not found
smartstring...not found
filesystem-rs...not found
conrod...found
generational-arena...found
yaml-rust...not found
not found (87)
    term                     https://github.com/rustsec/advisory-db/issues/134 
    termcolor                https://github.com/rustsec/advisory-db/issues/173 
    term                     https://github.com/rustsec/advisory-db/issues/173 
    rusb                     https://github.com/rustsec/advisory-db/issues/173 
    ring                     https://github.com/rustsec/advisory-db/issues/173 
    crossbeam                https://github.com/rustsec/advisory-db/issues/173 
    libusb-rs                https://github.com/rustsec/advisory-db/issues/173 
    rusb                     https://github.com/rustsec/advisory-db/issues/180 
    libusb-rs                https://github.com/rustsec/advisory-db/issues/180 
    term                     https://github.com/rustsec/advisory-db/issues/182 
    cassandra-rs             https://github.com/rustsec/advisory-db/issues/183 
    quickersort              https://github.com/rustsec/advisory-db/issues/210 
    xdg                      https://github.com/rustsec/advisory-db/issues/282 
    home                     https://github.com/rustsec/advisory-db/issues/282 
    directories-rs           https://github.com/rustsec/advisory-db/issues/282 
    dirs-rs                  https://github.com/rustsec/advisory-db/issues/282 
    dirs-sys-rs              https://github.com/rustsec/advisory-db/issues/282 
    traits                   https://github.com/rustsec/advisory-db/issues/310 
    rust-xcb                 https://github.com/rustsec/advisory-db/issues/653 
    crate                    https://github.com/rustsec/advisory-db/issues/742 
    calamine                 https://github.com/rustsec/advisory-db/issues/742 
    ab-glyph                 https://github.com/rustsec/advisory-db/issues/867 
    anymap                   https://github.com/rustsec/advisory-db/issues/906 
    cosmos-rust              https://github.com/rustsec/advisory-db/issues/1010
    destructure_traitobject  https://github.com/rustsec/advisory-db/issues/1088
    rust-traitobject         https://github.com/rustsec/advisory-db/issues/1088
    dryoc                    https://github.com/rustsec/advisory-db/issues/1090
    AEADs                    https://github.com/rustsec/advisory-db/issues/1090
    libsodium-sys-stable     https://github.com/rustsec/advisory-db/issues/1090
    rust-ansi-term           https://github.com/rustsec/advisory-db/issues/1095
    windows-rs               https://github.com/rustsec/advisory-db/issues/1098
    xml-rs                   https://github.com/rustsec/advisory-db/issues/1121
    cargo-download           https://github.com/rustsec/advisory-db/issues/1132
    prettytable-rs           https://github.com/rustsec/advisory-db/issues/1173
    owning-ref-unsoundness   https://github.com/rustsec/advisory-db/issues/1174
    futures-intrusive        https://github.com/rustsec/advisory-db/issues/1175
    users                    https://github.com/rustsec/advisory-db/issues/1179
    markdown                 https://github.com/rustsec/advisory-db/issues/1191
    maligned                 https://github.com/rustsec/advisory-db/issues/1209
    chrono                   https://github.com/rustsec/advisory-db/issues/1216
    toml-rs                  https://github.com/rustsec/advisory-db/issues/1221
    grass                    https://github.com/rustsec/advisory-db/issues/1228
    rsass                    https://github.com/rustsec/advisory-db/issues/1228
    erased_set               https://github.com/rustsec/advisory-db/issues/1236
    double-checked-cell      https://github.com/rustsec/advisory-db/issues/1239
    dotenv                   https://github.com/rustsec/advisory-db/issues/1254
    anymap                   https://github.com/rustsec/advisory-db/issues/1281
    interledger-rs           https://github.com/rustsec/advisory-db/issues/1303
    serde-xml-rs             https://github.com/rustsec/advisory-db/issues/1357
    interledger-rs           https://github.com/rustsec/advisory-db/issues/1369
    anymap                   https://github.com/rustsec/advisory-db/issues/1378
    docs                     https://github.com/rustsec/advisory-db/issues/1401
    num-format               https://github.com/rustsec/advisory-db/issues/1405
    ttmap                    https://github.com/rustsec/advisory-db/issues/1406
    typemap                  https://github.com/rustsec/advisory-db/issues/1406
    typemap_rev              https://github.com/rustsec/advisory-db/issues/1406
    rust-typemap             https://github.com/rustsec/advisory-db/issues/1406
    num-format               https://github.com/rustsec/advisory-db/issues/1416
    wyz                      https://github.com/rustsec/advisory-db/issues/1447
    surf                     https://github.com/rustsec/advisory-db/issues/1471
    hyper                    https://github.com/rustsec/advisory-db/issues/1482
    rustc-hex                https://github.com/rustsec/advisory-db/issues/1514
    rust-claim               https://github.com/rustsec/advisory-db/issues/1521
    json                     https://github.com/rustsec/advisory-db/issues/1536
    json-rust                https://github.com/rustsec/advisory-db/issues/1536
    daemonize                https://github.com/rustsec/advisory-db/issues/1543
    rust-boxfnonce           https://github.com/rustsec/advisory-db/issues/1603
    mach2                    https://github.com/rustsec/advisory-db/issues/1604
    mach                     https://github.com/rustsec/advisory-db/issues/1604
    rust-encoding            https://github.com/rustsec/advisory-db/issues/1605
    tui-rs                   https://github.com/rustsec/advisory-db/issues/1610
    tui-rs-revival           https://github.com/rustsec/advisory-db/issues/1610
    mbox                     https://github.com/rustsec/advisory-db/issues/1631
    malloc_buf               https://github.com/rustsec/advisory-db/issues/1635
    AEADs                    https://github.com/rustsec/advisory-db/issues/1694
    sysinfo                  https://github.com/rustsec/advisory-db/issues/1701
    rust-users               https://github.com/rustsec/advisory-db/issues/1701
    dlopen2                  https://github.com/rustsec/advisory-db/issues/1735
    quote                    https://github.com/rustsec/advisory-db/issues/1735
    syn                      https://github.com/rustsec/advisory-db/issues/1735
    teams                    https://github.com/rustsec/advisory-db/issues/1756
    hpack-patched            https://github.com/rustsec/advisory-db/issues/1784
    loopdev                  https://github.com/rustsec/advisory-db/issues/1821
    rust-cpython             https://github.com/rustsec/advisory-db/issues/1822
    smartstring              https://github.com/rustsec/advisory-db/issues/1854
    filesystem-rs            https://github.com/rustsec/advisory-db/issues/1870
    yaml-rust                https://github.com/rustsec/advisory-db/issues/1921
found (41)
    chan                     https://github.com/rustsec/advisory-db/issues/173 
    rust-crypto              https://github.com/rustsec/advisory-db/issues/173 
    sodiumoxide              https://github.com/rustsec/advisory-db/issues/173 
    chan                     https://github.com/rustsec/advisory-db/issues/179 
    rust-crypto              https://github.com/rustsec/advisory-db/issues/181 
    rust_sodium              https://github.com/rustsec/advisory-db/issues/225 
    failure                  https://github.com/rustsec/advisory-db/issues/368 
    stderr                   https://github.com/rustsec/advisory-db/issues/585 
    office                   https://github.com/rustsec/advisory-db/issues/742 
    rulinalg                 https://github.com/rustsec/advisory-db/issues/801 
    block-ciphers            https://github.com/rustsec/advisory-db/issues/893 
    block-ciphers            https://github.com/rustsec/advisory-db/issues/895 
    block-ciphers            https://github.com/rustsec/advisory-db/issues/897 
    utils                    https://github.com/rustsec/advisory-db/issues/904 
    nacl-compat              https://github.com/rustsec/advisory-db/issues/1090
    sodiumoxide              https://github.com/rustsec/advisory-db/issues/1090
    serde_cbor               https://github.com/rustsec/advisory-db/issues/1110
    cbor                     https://github.com/rustsec/advisory-db/issues/1110
    lmdb-rkv                 https://github.com/rustsec/advisory-db/issues/1142
    lmdb-rs                  https://github.com/rustsec/advisory-db/issues/1142
    sass-rs                  https://github.com/rustsec/advisory-db/issues/1228
    static_type_map          https://github.com/rustsec/advisory-db/issues/1236
    wee_alloc                https://github.com/rustsec/advisory-db/issues/1394
    badge                    https://github.com/rustsec/advisory-db/issues/1401
    orbtk                    https://github.com/rustsec/advisory-db/issues/1436
    multipart                https://github.com/rustsec/advisory-db/issues/1438
    parity-wasm              https://github.com/rustsec/advisory-db/issues/1441
    http-types               https://github.com/rustsec/advisory-db/issues/1471
    rusoto                   https://github.com/rustsec/advisory-db/issues/1494
    buf_redux                https://github.com/rustsec/advisory-db/issues/1602
    kuchiki                  https://github.com/rustsec/advisory-db/issues/1611
    buf_redux                https://github.com/rustsec/advisory-db/issues/1614
    wither                   https://github.com/rustsec/advisory-db/issues/1630
    crypto-hash              https://github.com/rustsec/advisory-db/issues/1633
    io_ext                   https://github.com/rustsec/advisory-db/issues/1706
    cloud-storage            https://github.com/rustsec/advisory-db/issues/1715
    cloud-storage-rs         https://github.com/rustsec/advisory-db/issues/1715
    hpack                    https://github.com/rustsec/advisory-db/issues/1784
    loopdev-erikh            https://github.com/rustsec/advisory-db/issues/1821
    conrod                   https://github.com/rustsec/advisory-db/issues/1871
    generational-arena       https://github.com/rustsec/advisory-db/issues/1883
